In [2]:
import numpy as np
import pylab as plt

In [6]:
alpha=0.05

P_0 = 3.5*0.01 
P_1 = 70.1*0.01

n_0 = 409
n_1 = 564

Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

In [9]:
sp_stdev = lambda p, n: np.sqrt(p * (1 - p) / n[0] +  p * (1 - p) / n[1])

sp_stdev_2y = sp_stdev(( Nt_0 + Nt_1) / (n_0 + n_1), [n_0, n_1])
print P_0, n_0, n_1, sp_stdev_2y

0.035 409 564 0.0320658086057


In [11]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, sp_stdev_2y)
print z_2y

20.7697865408


![title](ztable.jpg)

In [33]:
p_2y = 1 - 0.9997

print 'Since the p-value', p_2y, 'is less than the critical value 0.05, we have enough evidence to reject the null hypothesis'

Since the p-value 0.0003 is less than the critical value 0.05, we have enough evidence to reject the null hypothesis


## What if we used the values for where the former inmate was or was not "Convicted of a felony" (row 10) in the Recidivism (Years 1-3)?

![title](felony.png)

## Null hypothesis: The % of former prisoners convicted again after release is the same or higher for candidates who participated in the program compared to the control group 

## Significance level: a=0.05

Assume that $P = y / n$ is the proportion of prisoners convicted again, with $y$ being the number of prisoners convicted and $n$ the sample size.

## $H_0: P_0 - P_1 \leq 0$
## $H_a: P_0 - P_1 > 0$

## $P_0 = 0.117, P_1= 0.100$

In [48]:
alpha = 0.05

P_0 = 0.117 
P_1 = 0.1

n_0 = 409
n_1 = 568

count_0 = P_0 * n_0
count_1 = P_1 * n_1

p_hat = (count_0 + count_1)/(n_0 + n_1)

In [49]:
sp_stdev = lambda p, n1, n2: np.sqrt(p * (1 - p) / n1 +  p * (1 - p) / n2)

sp_stdev_3y = sp_stdev(p_hat, n_0, n_1)
print P_0, n_0, n_1, sp_stdev_3y

0.117 409 568 0.0200556791612


In [50]:
zscore = lambda p_0, p_1, se : (p_0 - p_1) / se
z_3y = zscore(P_1, P_0, sp_stdev_2y)
print z_3y

-0.84764020522


In [51]:
pvalue = 0.1711+((-0.847+0.9)*(0.1911-0.1711)/(-0.8+0.9)) # Interpolation from z-table 0.8 and 0.9 values
pvalue

0.1817

In [53]:
p_3y = 1 - 0.1817

print 'Since the p-value', p_3y, 'is higher than the critical value 0.05, we fail to reject the null hypothesis'

Since the p-value 0.8183 is higher than the critical value 0.05, we fail to reject the null hypothesis


# Now lets do it with the $\chi^2$ test

In [54]:
chisqstat= lambda N, values, expect_num : ((values[0][0] * values[1][1]
                                            -values[0][1] * values[1][0])**2) * N / expect_num
Ntot = 973  # a + b + c + d = tot
expected_num = 564 * 409 * 409.679 * 563.321
sample_values = [[0.701 * 564, 0.299 * 564], [0.0305 * 409, 0.965 * 409]]
 

print (chisqstat(Ntot,  sample_values, expected_num))

433.128818315


When comparing two samples the degrees of freedom are derived from the following formula:

(Number of rows - 1) x (Number of columns -1) in the contigency table (https://en.wikipedia.org/wiki/Chi-squared_test)

Thus, the we have __df=1__ for this problem

![title](chi-square-table.gif)

For __a = 0.05__, the value of the chi square table is __3.84__, which is way smaller than 433.1. Thus, we reject $H_{0}$ that the percentage of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group with 95% confidence.

The result is consistent with the previous findings of the __z test__.

The contigency table for conviction for felonies is as follows:


|convicted of a fellony     |     employed   | not employes   | total  |                  
|---------------------------|----------------|----------------|--------|
| test sample               | 0.1x568            |   0.9x568             |    568    |          
| control sample            | 0.117x409               |  0.883x409              |    409    |                    
| __total__                 | 104.653               |  872.347              |    977    |

In [75]:
table = [[0.1 * 568, 0.9 * 568], [0.117 * 409, 0.883 * 409]]
prod_tot = 104.653*872.347*568*409
N = 977

chi2stat = ((table[0][0]*table[1][1]-table[0][1]*table[1][0])**2) * N / prod_tot
print 'The chisquared statistic is:' ,chi2stat

The chisquared statistic is: 0.718493917505


For __a=.05__ and for __df=1__ (similarly to the previous example), the value is __3.84__, which is higher than the $\chi^{2}=0.781$ we obtained. So, we fail to reject the null hypothesis that the percentage of former prisoners convicted again after release is the same or higher for candidates who participated in the program compared to the control group.

Again, the finding is consistent with the __z test__.